server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java

```diff
@@ -97,6 +108,25 @@ public void handle(Request request, Response response) throws Exception {
     response.noContent();
   }
 
+  private void validate(SetRequest request, Optional<ComponentDto> component) {
+    PropertyDefinition definition = propertyDefinitions.get(request.getKey());
+    if (definition == null) {
+      return;
+    }
+
+    PropertyDefinition.Result result = definition.validate(request.getValue());
+
+    checkRequest(result.isValid(),
+      i18n.message(Locale.ENGLISH, "property.error." + result.getErrorKey(), "Error when validating setting with key '%s' and value '%s'"),
+      request.getKey(), request.getValue());
+
+    checkRequest(component.isPresent() || definition.global(), "Setting '%s' cannot be global", request.getKey());
+    String qualifier = component.isPresent() ? component.get().qualifier() : "";
+    checkRequest(!component.isPresent()
+      || definition.qualifiers().contains(component.get().qualifier()),
+      "Setting '%s' cannot be set on a component with qualifier '%s'", request.getKey(), qualifier);
```

Comment:
I'm not sure that displaying the qualifier from DB is a good idea... Will "TRK" talk to a user ?
Why not displaying the value from the bundle, for instance "qualifier.TRK" => "Project".
The message should then be changed to "Setting '%s' cannot be set on a '%s'".